api/types/access_request.go

```diff
@@ -185,6 +206,255 @@ func (r *AccessRequestV3) SetSystemAnnotations(annotations map[string][]string)
+	// and aggregate the results.
+	for _, role := range roles {
+		thresholdSets, ok := r.Spec.RoleThresholdMapping[role]
+		if !ok {
+			return trace.BadParameter("role %q is not a member of this request", role)
+		}
+
+		for _, tset := range thresholdSets.Sets {
+			for _, tid := range tset.Indexes {
+				if _, ok := matches[tid]; ok {
+					// already checked this threshold
+					continue
+				}
+
+				thresh, err := r.getThreshold(tid)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				match, err := thresh.MatchesFilter(parser)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+
+				matches[tid] = match
+				if match {
+					matchCount++
+				}
+			}
+		}
+	}
+
+	// record a list of all thresholds whose filters match this review.
+	rev.ThresholdIndexes = make([]uint32, 0, matchCount)
+	for tid, matched := range matches {
+		if matched {
+			rev.ThresholdIndexes = append(rev.ThresholdIndexes, tid)
+		}
+	}
+
+	r.Spec.Reviews = append(r.Spec.Reviews, rev)
+
+	// recalculate updated request state.
+	return r.onReview()
+}
+
+// onReview checks if we've hit sufficient thresholds for a state-transition,
+// and applies it if that is the case.
+func (r *AccessRequestV3) onReview() error {
+	if !r.Spec.State.IsPending() {
+		// no further state-transitions are performed after we exit
+		// the PENDING state.
+		return nil
+	}
+
+	// TODO(fspmarshall): Rework this function to support role subselection.
+
+	// approved keeps track of roles that have hit at least one
+	// of their approval thresholds.
+	approved := make(map[string]bool)
+	// counts keeps track of the approval and denial counts for all thresholds.
+	counts := make([]struct{ approval, denial uint32 }, len(r.Spec.Thresholds))
+
+	// lastReview stores the most recently processed review.  Because processing
+	// halts once approval is reached, this represents the Nth review where N is
+	// the highest applicable approval threshold.
+	var lastReview AccessReview
+
+	// Iterate through all reviews and aggregate them against `counts`.
+ProcessReviews:
+	for _, rev := range r.Spec.Reviews {
+		lastReview = rev
+		for _, tid := range rev.ThresholdIndexes {
+			idx := int(tid)
+			if len(r.Spec.Thresholds) <= idx {
+				return trace.Errorf("threshold index '%d' out of range (this is a bug)", idx)
+			}
+			if rev.State.IsApproved() {
+				counts[idx].approval++
+			}
+			if rev.State.IsDenied() {
+				counts[idx].denial++
+			}
+		}
+
+		// If we hit any denial thresholds, short-circuit immediately
+		for i, t := range r.Spec.Thresholds {
+
+			if counts[i].denial >= t.Deny && t.Deny != 0 {
+				// A single denial threshold has been met, deny entire request
+				r.Spec.State = RequestState_DENIED
+				r.Spec.ResolveReason = lastReview.Reason
+				r.Spec.ResolveAnnotations = lastReview.Annotations
+				return nil
+			}
+		}
+
+		// check for roles that can be transitioned to an approved state
+	CheckRoleApprovals:
```

Comment:
You're not missing anything, but IMO the current layout is superior because it is deterministic regardless of the length of the review list.  If we tally all votes and then determined the state-transition as separate steps, then the result that would be produced by this logic continues to change as new reviews are added, even after the initial state-transition condition is reached.

Currently, this property is irrelevant since we skip this entire function once we've exited the `PENDING` state, but the fact that the function is deterministic means that it will be much easier to extend in the future.  Say that we wanted to add an integration with the upcoming "locking" feature which automatically generated locks if a denial theshold was triggered *after* an approval was initially triggered.  The current logic could be left basically untouched, and we could just add a new loop below that starts processing where the previous loop halted.  Basically, "its not a bug, its a feature" lol.